### PR TITLE
Add state-in-constructor rule

### LINKED ISSUE #1810: Rule suggestion: no-state-in-constructor
**INCORRECT**
```javascript
class MyComponent extends Component {
  constructor(props) {
    super(props);

    this.state = {
      counter: props.initialValue
    };
  }
}
```

**CORRECT**
```javascript
class MyComponent extends Component {
  state = {
    counter: this.props.initialValue
  };
}
```

**Why?**
It may reduce boilerplate by not having to define a constructor just for initializing the state. 

What do you think about this?

### PR DESCRIPTION
**Pull request by lukyth**
This will resolve #1810. Comments are very welcomed. 😃

### CODE REVIEW

File: lib/rules/state-in-constructor.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            message: 'State initialization should be in a constructor'
+          });
+        }
+      },
+      AssignmentExpression(node) {
+        if (
+          option === 'never' &&
+          node.left.object.type === 'ThisExpression' &&
+          node.left.property.name === 'state' &&
+          node.parent.parent.parent.parent.kind === 'constructor' &&
```
Comment (alexzherdev) [inline lib/rules/state-in-constructor.js]: This is a bit brittle: assignment could be nested in other nodes (e.g. if conditions) in which case this sequence will not work. This should either be traversing parents in a loop, or scopes like https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/no-unused-prop-types.js#L264 (although not sure which one is better)

File: lib/rules/state-in-constructor.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          utils.isES6Component(node.parent.parent.parent.parent.parent.parent)
```
Comment (alexzherdev) [inline lib/rules/state-in-constructor.js]: Same here, but in this case maybe `utils.getParentES6Component` would work

File: lib/rules/state-in-constructor.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          context.report({
+            node,
+            message: 'State initialization should be in a constructor'
+          });
+        }
+      },
+      AssignmentExpression(node) {
+        if (
+          option === 'never' &&
+          node.left.object.type === 'ThisExpression' &&
```
[the remark this answers is not in the document]
Comment (ljharb) [reply lib/rules/state-in-constructor.js]: Let's try to add a test case that would fail here.

File: lib/rules/state-in-constructor.js (hunk 1 above)
Comment (ljharb) [reply lib/rules/state-in-constructor.js]: Let's try to add a test case that would fail here.

**Comment by lukyth** [comment]
@alexzherdev @ljharb Thank you so much for your reviews! I've addressed all of them with these commits 81d0e8b...b1024c0. Could you please check them out?

I also removed some test cases which I think are redundant (81d0e8b). I'm not sure if that's the right call. If you guys don't think so then I can bring them back.

**Comment by krawaller** [comment]
Just to inject some energy, this looks sweet! Good work @lukyth! :)

**Comment by lukyth** [comment]
Anything I can do to get this merged?

**Comment by lukyth** [comment]
@EvHaus I've added this rule to the list in `README.md` (9415814). Sorry it took me so long. I just saw your comment here.

**Pull request merged**